### PR TITLE
Send remote access headers for internal users

### PR DESCRIPTION
**Pull request by n1v0lg**
This PR adds querying-cluster-side support for internal users making requests cross cluster under the new security model. 

For the querying cluster, this involves detecting that a request is made by an internal user and sending a remote access header with an empty role descriptors intersection and an embedded authentication instance identifying the internal user making the request. The fulfilling cluster will use this header to resolve the correct privileges for the given internal user (while also enforcing the privileges of the remote access key used for the request). The fulfilling cluster side implementation will follow in a separate PR.

Currently, the only use of internal users cross cluster is requests made by `_system` to fetch cluster-state from the fulfilling cluster.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/transport/RemoteConnectionManager.java
Comment (n1v0lg) [inline server/src/main/java/org/elasticsearch/transport/RemoteConnectionManager.java:86]: Separate PR open for this here, which will need to merge first: https://github.com/elastic/elasticsearch/pull/92584

Comment (ywangd) [reply server/src/main/java/org/elasticsearch/transport/RemoteConnectionManager.java:86]: Does this PR depend on this change?

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java
Comment (ywangd) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java:309]: I don't think this change is intended?

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java
Comment (ywangd) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java:354]: I have two suggestions for alternatives for your consideration:
1. Move this logic of using empty roleDescriptors into `AuthorizationService` so that this method here does not have to change. I can see an argument for keeping ther interalUser check closer to where request is sent. But any request sent to remote cluster needs to retrieve remote privileges. So handling it in `AuthorizationService#retrieveRemoteAccessRoleDescriptorsIntersection` can actually help it fail earlier on. It is deeper in the callstack which means another caller (if any in future) does not have to handle the special logic on its own. Concretely this means, returning empty for `_system` and throw for all other internal users inside `AuthorizationService#retrieveRemoteAccessRoleDescriptorsIntersection`.
2. Handle `_system` more explicitly with something like `if (SystemUser.is()) { ... } else if (Users.isInternal) { ... } else { ... }`. This is mainly to remove `assertInternalUserExpectedInCrossClusterRequest` which I found a bit implicit.

I slightly prefer 1.

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptorTests.java
Comment (ywangd) [inline x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptorTests.java:563]: Do we have a test for throwing exceptions on other internal users?

**Review by ywangd — APPROVED** [review]
LGTM with a few minor comments 

Also, if this PR does have to depend on #92584, please wait till #92584 is merged and incorporate upstream change before merge. Thanks!

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java
Comment (n1v0lg) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java:309]: Whoops! Thanks for catching

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java
Comment (n1v0lg) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptor.java:354]: We discussed on Slack: going with option 2 since the semantics of `retrieveRemoteAccessRoleDescriptorsIntersection` can be argued to be clearer if the method simply doesn't handle internal users and because special-casing the system user is slightly more consistent with how the system user is handled elsewhere (since the system user does not an actual role, it also makes sense that it's not supported to retrieve any kind of role descriptor for it).

File: x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptorTests.java
Comment (n1v0lg) [reply x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/transport/SecurityServerTransportInterceptorTests.java:563]: No test, but we have an assertion in the interceptor that covers other internal users. To test this I'd have to toggle off the assertion which feels off. I'll leave as is for now, but I can instead get rid of the assertion and add a test -- let me know if you prefer this (there is still time since this PR is blocked anyways).

File: server/src/main/java/org/elasticsearch/transport/RemoteConnectionManager.java
Comment (n1v0lg) [reply server/src/main/java/org/elasticsearch/transport/RemoteConnectionManager.java:86]: Yup this PR depends on https://github.com/elastic/elasticsearch/pull/92584 (since there are scenarios where a connection is actually remote that we don't currently detect -- the linked PR addresses this) -- I will wait for approval on 92584, merge that, then merge this current PR.

**Comment by n1v0lg** [comment]
@elasticsearchmachine run elasticsearch-ci/part-1-fips

**Pull request merged**
